File: Gabriel_Code/Prototipo_0_Java/App/AbstractGraph.java
package br.pucminas.grafo.graph;

/**
 * Classe abstrata que define a API obrigatória do trabalho prático.
 *
 * <p>Implementa atributos compartilhados (rótulos e pesos de vértices)
 * e métodos auxiliares de validação de índice. As subclasses fornecem
 * a representação interna (matriz ou lista de adjacência).</p>
 */
public abstract class AbstractGraph {

    protected final int numVertices;
    protected final double[] vertexWeights;
    protected final String[] vertexLabels;

    /** Tipo de representação interna, para exibição e export. */
    public enum RepType { MATRIX, LIST }
    protected final RepType repType;

    // ── Construtor ─────────────────────────────────────────────────────────

    protected AbstractGraph(int numVertices, RepType repType) {
        if (numVertices <= 0)
            throw new IllegalArgumentException("numVertices deve ser > 0.");
        this.numVertices  = numVertices;
        this.repType      = repType;
        this.vertexWeights = new double[numVertices];
        this.vertexLabels  = new String[numVertices];
        for (int i = 0; i < numVertices; i++) {
            vertexWeights[i] = 1.0;
            vertexLabels[i]  = "v" + i;
        }
    }

    // ── Validação ──────────────────────────────────────────────────────────

    protected void checkVertex(int v) {
        if (v < 0 || v >= numVertices)
            throw new IndexOutOfBoundsException(
                "Vértice inválido: " + v + " (total=" + numVertices + ")");
    }

    protected void checkEdge(int u, int v) {
        checkVertex(u);
        checkVertex(v);
        if (u == v) throw new IllegalArgumentException("Laço não permitido: " + u);
    }

    // ── API obrigatória ────────────────────────────────────────────────────

    public int getVertexCount() { return numVertices; }

    public abstract int getEdgeCount();

    public abstract boolean hasEdge(int u, int v);

    public abstract void addEdge(int u, int v);

    public abstract void removeEdge(int u, int v);

    /** v é sucessor de u (existe aresta u→v)? */
    public boolean isSuccessor(int u, int v)   { return hasEdge(u, v); }

    /** v é predecessor de u (existe aresta v→u)? */
    public boolean isPredecessor(int u, int v) { return hasEdge(v, u); }

    /**
     * As arestas (u1,v1) e (u2,v2) são divergentes?
     * Dois arcos são divergentes quando partem do mesmo vértice.
     */
    public boolean isDivergent(int u1, int v1, int u2, int v2) {
        checkEdge(u1, v1); checkEdge(u2, v2);
        return u1 == u2 && hasEdge(u1, v1) && hasEdge(u2, v2);
    }

    /**
     * As arestas (u1,v1) e (u2,v2) são convergentes?
     * Dois arcos convergem quando chegam ao mesmo vértice.
     */
    public boolean isConvergent(int u1, int v1, int u2, int v2) {
        checkEdge(u1, v1); checkEdge(u2, v2);
        return v1 == v2 && hasEdge(u1, v1) && hasEdge(u2, v2);
    }

    /**
     * A aresta (u,v) é incidente ao vértice x?
     * (x == u ou x == v, e a aresta existe)
     */
    public boolean isIncident(int u, int v, int x) {
        checkEdge(u, v); checkVertex(x);
        return (x == u || x == v) && hasEdge(u, v);
    }

    public abstract int getVertexInDegree(int u);

    public abstract int getVertexOutDegree(int u);

    // ── Pesos ──────────────────────────────────────────────────────────────

    public void setVertexWeight(int v, double w) {
        checkVertex(v);
        vertexWeights[v] = w;
    }

    public double getVertexWeight(int v) {
        checkVertex(v);
        return vertexWeights[v];
    }

    public void setVertexLabel(int v, String label) {
        checkVertex(v);
        vertexLabels[v] = label;
    }

    public String getVertexLabel(int v) {
        checkVertex(v);
        return vertexLabels[v];
    }

    public abstract void   setEdgeWeight(int u, int v, double w);
    public abstract double getEdgeWeight(int u, int v);

    // ── Propriedades estruturais ────────────────────────────────────────────

    /**
     * Verifica conectividade fraca: ignora direção das arestas.
     * Usa BFS no grafo subjacente não-direcionado.
     */
    public boolean isConnected() {
        if (numVertices == 0) return true;
        boolean[] visited = new boolean[numVertices];
        java.util.Queue<Integer> queue = new java.util.LinkedList<>();
        queue.add(0);
        visited[0] = true;
        int count = 1;
        while (!queue.isEmpty()) {
            int cur = queue.poll();
            for (int w = 0; w < numVertices; w++) {
                if (!visited[w] && (hasEdge(cur, w) || hasEdge(w, cur))) {
                    visited[w] = true;
                    queue.add(w);
                    count++;
                }
            }
        }
        return count == numVertices;
    }

    /** Grafo vazio: sem arestas. */
    public boolean isEmptyGraph() { return getEdgeCount() == 0; }

    /**
     * Grafo completo (torneio completo para grafos direcionados):
     * para cada par (i,j) com i≠j existe aresta i→j e j→i.
     */
    public boolean isCompleteGraph() {
        for (int i = 0; i < numVertices; i++)
            for (int j = 0; j < numVertices; j++)
                if (i != j && !hasEdge(i, j)) return false;
        return true;
    }

    // ── Export ─────────────────────────────────────────────────────────────

    public abstract void exportToGEPHI(String path);

    // ── Utilitários ────────────────────────────────────────────────────────

    public RepType getRepType() { return repType; }

    @Override
    public String toString() {
        return getClass().getSimpleName() +
               "[V=" + numVertices + ", E=" + getEdgeCount() + "]";
    }
}

File: Projeto_Python_Apendice/grafo/graph/abstract_graph.py
from abc import ABC, abstractmethod
from enum import Enum, auto
import collections

class RepType(Enum):
    MATRIX = auto()
    LIST = auto()

class AbstractGraph(ABC):
    def __init__(self, num_vertices: int, rep_type: RepType):
        if num_vertices <= 0:
            raise ValueError("num_vertices deve ser > 0.")
        self.num_vertices = num_vertices
        self.rep_type = rep_type
        self.vertex_weights = [1.0] * num_vertices
        self.vertex_labels = [f"v{i}" for i in range(num_vertices)]

    def check_vertex(self, v: int):
        if v < 0 or v >= self.num_vertices:
            raise IndexError(f"Vértice inválido: {v} (total={self.num_vertices})")

    def check_edge(self, u: int, v: int):
        self.check_vertex(u)
        self.check_vertex(v)
        if u == v:
            raise ValueError(f"Laço não permitido: {u}")

    def get_vertex_count(self) -> int:
        return self.num_vertices

    @abstractmethod
    def get_edge_count(self) -> int:
        pass

    @abstractmethod
    def has_edge(self, u: int, v: int) -> bool:
        pass

    @abstractmethod
    def add_edge(self, u: int, v: int):
        pass

    @abstractmethod
    def remove_edge(self, u: int, v: int):
        pass

    def is_successor(self, u: int, v: int) -> bool:
        return self.has_edge(u, v)

    def is_predecessor(self, u: int, v: int) -> bool:
        return self.has_edge(v, u)

    def is_divergent(self, u1: int, v1: int, u2: int, v2: int) -> bool:
        self.check_edge(u1, v1)
        self.check_edge(u2, v2)
        return u1 == u2 and self.has_edge(u1, v1) and self.has_edge(u2, v2)

    def is_convergent(self, u1: int, v1: int, u2: int, v2: int) -> bool:
        self.check_edge(u1, v1)
        self.check_edge(u2, v2)
        return v1 == v2 and self.has_edge(u1, v1) and self.has_edge(u2, v2)

    def is_incident(self, u: int, v: int, x: int) -> bool:
        self.check_edge(u, v)
        self.check_vertex(x)
        return (x == u or x == v) and self.has_edge(u, v)

    @abstractmethod
    def get_vertex_in_degree(self, u: int) -> int:
        pass

    @abstractmethod
    def get_vertex_out_degree(self, u: int) -> int:
        pass

    def set_vertex_weight(self, v: int, w: float):
        self.check_vertex(v)
        self.vertex_weights[v] = w

    def get_vertex_weight(self, v: int) -> float:
        self.check_vertex(v)
        return self.vertex_weights[v]

    def set_vertex_label(self, v: int, label: str):
        self.check_vertex(v)
        self.vertex_labels[v] = label

    def get_vertex_label(self, v: int) -> str:
        self.check_vertex(v)
        return self.vertex_labels[v]

    @abstractmethod
    def set_edge_weight(self, u: int, v: int, w: float):
        pass

    @abstractmethod
    def get_edge_weight(self, u: int, v: int) -> float:
        pass

    def is_connected(self) -> bool:
        if self.num_vertices == 0:
            return True
        visited = [False] * self.num_vertices
        queue = collections.deque([0])
        visited[0] = True
        count = 1
        while queue:
            cur = queue.popleft()
            for w in range(self.num_vertices):
                if not visited[w] and (self.has_edge(cur, w) or self.has_edge(w, cur)):
                    visited[w] = True
                    queue.append(w)
                    count += 1
        return count == self.num_vertices

    def is_empty_graph(self) -> bool:
        return self.get_edge_count() == 0

    def is_complete_graph(self) -> bool:
        for i in range(self.num_vertices):
            for j in range(self.num_vertices):
                if i != j and not self.has_edge(i, j):
                    return False
        return True

    @abstractmethod
    def export_to_gephi(self, path: str):
        pass

    def __str__(self):
        return f"{self.__class__.__name__}[V={self.num_vertices}, E={self.get_edge_count()}]"
